src/Agent.Listener/Agent.cs

```diff
@@ -137,6 +142,17 @@ public override void Initialize(IHostContext hostContext)
             finally
             {
                 _term.CancelKeyPress -= CtrlCHandler;
+                loadContext.Unloading -= Agent_Unloading;
+                _completedCommand.Set();
+            }
+        }
+
+        private void Agent_Unloading(AssemblyLoadContext obj)
+        {
+            if ((!_inConfigStage) && (!TokenSource.IsCancellationRequested))
+            {
+                TokenSource.Cancel();
+                _completedCommand.WaitOne(TimeSpan.FromSeconds(30));
```

Comment:
It would be good to put this constant into constants.cs. I just grepped the source code for "30", and found more timeouts than I expected. This specific timeout needs to be coordinated with the rest of the timeouts across processes (all the way down to the task timeout when that is implemented).